experimental/iterators/lib/Conversion/IteratorsToStandard/IteratorsToStandard.cpp

```diff
@@ -136,6 +191,104 @@ struct IteratorConversionPattern : public ConversionPattern {
+  ConstantTupleLowering(MLIRContext *context, PatternBenefit benefit = 1)
+      : ConversionPattern("iterators.constant", benefit, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create LLVM struct type.
+    assert(op->getNumResults() == 1);
+    TupleType resultType = op->getResult(0).getType().dyn_cast<TupleType>();
+    assert(resultType);
+    LLVMStructType structType = LLVMStructType::getNewIdentified(
+        getContext(), "tuple", resultType.getTypes());
+
+    // Undef.
+    Value structValue = rewriter.create<UndefOp>(op->getLoc(), structType);
+
+    // Insert values.
+    ArrayAttr values = op->getAttr("values").dyn_cast<ArrayAttr>();
+    assert(values);
+    for (int i = 0; i < static_cast<int>(values.size()); i++) {
+      // Create index attribute.
+      ArrayAttr indicesAttr = rewriter.getIndexArrayAttr({i});
+
+      // Create constant value op.
+      Value valueOp = rewriter.create<LLVM::ConstantOp>(
+          op->getLoc(), values[i].getType(), values[i]);
+
+      // Insert into struct.
+      structValue = rewriter.create<InsertValueOp>(op->getLoc(), structValue,
+                                                   valueOp, indicesAttr);
+    }
+
+    rewriter.replaceOp(op, structValue);
+
+    return success();
+  }
+};
+
+struct PrintOpLowering : public ConversionPattern {
+  PrintOpLowering(MLIRContext *context, PatternBenefit benefit = 1)
+      : ConversionPattern("iterators.print", benefit, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(op->getNumOperands() == 1);
+    TupleType tupleType = op->getOperand(0).getType().dyn_cast<TupleType>();
+    assert(tupleType);
+
+    assert(operands.size() == 1);
+    LLVMStructType structType =
+        operands[0].getType().dyn_cast<LLVMStructType>();
+    assert(structType);
+
+    // Assemble format string in the form `(%i, %i, ...)`.
+    SmallString<8> format("(");
+    for (auto const &type : tupleType.getTypes()) {
+      // TODO(ingomueller): add type switch for other number types.
+      assert(type == rewriter.getI32Type() && "Only I32 is supported for now");
+      format += "%i, ";
+    }
+    // Remove last ", " unless the format is empty
+    if (format.size() > 1) {
+      format.erase(format.end() - 2, format.end());
+    }
+    format += ")\n";
+    format.push_back('\0');
+
+    // Insert format string as global.
+    ModuleOp module = op->getParentOfType<ModuleOp>();
+    Value formatSpec = getOrCreateGlobalString(
+        rewriter, Twine("frmt_spec.") + structType.getName(), format, module);
+
+    // Assemble arguments to printf.
+    SmallVector<Value, 8> values = {formatSpec};
+    ArrayRef<Type> fieldTypes = structType.getBody();
+    for (int i = 0; i < static_cast<int>(fieldTypes.size()); i++) {
```

Comment:
seems like you could coiterate with line 251 (with llvm::zip)

Reply:
I need the value of `i` in the next line, though, as the index for `LLVMExtractValue`. With `zip`, I get a range without index, no?

Reply:
Obsolete due to `llvm::interleaveComma`.

Reply:
TODO: look into `enumerate`.

Reply:
I tried out `enumerate` and didn't find it better: it makes the use of both `i` and `fieldTypes[i]` longer (`it.index()` and `it.value()`) and it still requires a cast (because `IndexArrayAttr` cannot be constructed from an `std::size_t` without warning).